### PR TITLE
JTAG Flash Support (VSC-487)

### LINKED ISSUE #161: [Feature Request]: Upload firmware via jtag. (VSC-427)
I want to use your extension, but impossible upload firmware in my case.
My pcb has only jtag port.
No serial.....

Thank you!

### PR DESCRIPTION
**Pull request by pwmb**
This feature will enable to flash device via JTAG

##### TODOs:

- [x] Add `i18n`
- [x] Restrict simultaneous flash via `UART` & `JTAG`

Closes #161 

---

##### How to test?

- Set the settings `idf.flashType` as `JTAG` and click flash button and it should use the functionality for flashing via JTAG

---

##### Developers Note

> OpenOCD Version Requirements: [v0.10.0-esp32-20201125](https://github.com/espressif/openocd-esp32/releases/tag/v0.10.0-esp32-20201125) and above

> Awaits release and backport to IDF

### CODE REVIEW

**Comment by pwmb** [comment]
> Note: Must have OpenOCD version [v0.10.0-esp32-20201125](https://github.com/espressif/openocd-esp32/releases/tag/v0.10.0-esp32-20201125) or later

**Review by newblaque — APPROVED** [review]

**Review by newblaque — APPROVED** [review]

File: src/extension.ts [hunk 1]
```diff
@@ -1529,6 +1531,78 @@ export async function activate(context: vscode.ExtensionContext) {
         wsServer.close();
       });
   });
+  registerIDFCommand("espIdf.jtag_flash", () => {
+    PreCheck.perform([openFolderCheck], async () => {
+      const buildFolder = path.join(workspaceRoot.fsPath, "build");
+      if (!existsSync(buildFolder)) {
```
Comment (brianignacio5) [inline src/extension.ts:1537]: We should also check `.elf` is generated, build folder is still created on build failure.

File: package.json [hunk 2]
```diff
@@ -667,6 +676,16 @@
...
         "title": "%espIdf.getEspMdf.title%"
+      },
+      {
+        "command": "espIdf.jtag_flash",
+        "title": "%espIdf.jtag_flash.title%",
+        "category": "ESP-IDF"
+      },
+      {
+        "command": "espIdf.selectFlashMethodAndFlash",
+        "title": "Select Flash Device and Flash",
```
Comment (brianignacio5) [inline package.json:687]: Add Localization for `Select Flash Method` ? 

Should also rename old  flash method to Flash (UART)

File: src/extension.ts (hunk 1 above)
Comment (brianignacio5) [reply src/extension.ts:1537]: Wouldn't be better to use `pathExists` from path-extra for async instead of fs `existsSync` ?

**Review by brianignacio5 — CHANGES_REQUESTED** [review]
Tested on MacOS.  Feature looks really well and working properly.

Left some comments otherwise LGTM.

**Review by brianignacio5 — APPROVED** [review]
LGTM

**Pull request merged**